timelock/program/src/instruction.rs

```diff
@@ -0,0 +1,590 @@
+    Ping,
+
+    /// Executes a command in the timelock set.
+    ///
+    ///   0. `[writable]` Transaction account you wish to execute.
+    ///   1. `[writable]` Timelock state account.
+    ///   2. `[]` Program being invoked account
+    ///   3. `[]` Timelock set account.
+    ///   4. `[]` Timelock config
+    ///   5. `[]` Timelock program account pub key.
+    ///   6. `[]` Clock sysvar.
+    ///   7+ Any extra accounts that are part of the instruction, in order
+    Execute {
+        /// Number of extra accounts
+        number_of_extra_accounts: u8,
+    },
+
+    /// [Requires tokens of the Governance mint or Council mint depending on type of TimelockSet]
+    /// Deposits voting tokens to be used during the voting process in a timelock.
+    /// These tokens are removed from your account and can be returned by withdrawing
+    /// them from the timelock (but then you will miss the vote.)
+    ///
+    ///   0. `[writable]` Governance voting record account. See Vote docs for more detail.
+    ///   1. `[writable]` Initialized Voting account to hold your received voting tokens.
+    ///   2. `[writable]` User token account to deposit tokens from.
+    ///   3. `[writable]` Source holding account for timelock that will accept the tokens in escrow.
+    ///   4. `[writable]` Voting mint account.
+    ///   5. `[]` Timelock set account.
+    ///   6. `[]` Transfer authority
+    ///   7. `[]` Timelock program mint authority
+    ///   8. `[]` Timelock program account pub key.
+    ///   9. `[]` Token program account.
+    DepositSourceTokens {
+        /// How many voting tokens to deposit
+        voting_token_amount: u64,
+    },
+
+    /// [Requires voting tokens]
+    /// Withdraws voting tokens.
+    ///
+    ///   0. `[writable]` Governance voting record account. See Vote docs for more detail.
+    ///   1. `[writable]` Initialized Voting account from which to remove your voting tokens.
+    ///   2. `[writable]` Initialized Yes Voting account from which to remove your voting tokens.
+    ///   3. `[writable]` Initialized No Voting account from which to remove your voting tokens.
+    ///   4. `[writable]` User token account that you wish your actual tokens to be returned to.
+    ///   5. `[writable]` Source holding account owned by the timelock that will has the actual tokens in escrow.
+    ///   6. `[writable]` Initialized Yes Voting dump account owned by timelock set to which to send your voting tokens.
+    ///   7. `[writable]` Initialized No Voting dump account owned by timelock set to which to send your voting tokens.
+    ///   8. `[writable]` Voting mint account.
+    ///   9. `[writable]` Yes Voting mint account.
+    ///   10. `[writable]` No Voting mint account.
+    ///   11. `[]` Timelock state account.
+    ///   12. `[]` Timelock set account.
+    ///   13. `[]` Transfer authority
+    ///   14. `[]` Yes Transfer authority
+    ///   15. `[]` No Transfer authority
```

Comment:
1 transfer authority should probably be enough

Reply:
No, you actually need one for each, I have found out. The transfer authority expires after one use. It was very unfortunate.

Reply:
What do you mean it expires?  If I create ephemeral key K, delegate my A tokens to K, and then delegate my B tokens to K, and then sign with K and provide my A and B token accounts, you're saying K can't transfer both in one instruction?

Reply:
Sorry, I actually was wrong up there. I was looking at this yesterday and realized I misremembered the cause of this particular error that caused me to do this. These authorities really are different keys. Each of these are authorities to transfer from DIFFERENT accounts. That's why they are different keys.

Reply:
Following on my comment about `approve`, since transfer authorities are typically keypairs that have been `approve`d to move some funds, you could `approve` your YES / NO / governance tokens to the same ephemeral keypair K, sign with that, and pass just that one in.  Does that make sense?

Reply:
That hadn't occurred to me. Now that I understand it better, it could. Yes, I'll make a note of that!

Reply:
Oh I think the reason it happened this way is the front end has a helper that sort of produces one key per call. I'd need to make my own custom token helper to pull that off, but that's ok.

Reply:
Ah gotcha, I wasn't aware of that.  Up to you then! On the other hand, every pubkey that you save gives you a bit more space on your stack and transaction

Reply:
No, I'm doing it your way. I like the idea of cutting these down. It always made me feel dirty. I'm just going to adjust the token helper, it should really have an option to pass your own key in instead of always making its own every time.

Reply:
Done